### PR TITLE
Ensure that the RandomFourierFeatures layer can be trained in eager mode

### LINKED ISSUE #39088: It is not possible to train the trainable parameters of the RandomFourierFeatures keras layer in eager mode
<em>Please make sure that this is a bug. As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes; minimal working example provided
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux-5.3.0-46-generic-x86_64-with-Ubuntu-18.04-bionic
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): 2.2.0.dev20200501
- Python version: 3.7.5
- Bazel version (if compiling from source): N/A
- GCC/Compiler version (if compiling from source): N/A
- CUDA/cuDNN version: N/A
- GPU model and memory: N/A

You can collect some of this information using our environment capture
[script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with:
1. TF 1.0: `python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"`
2. TF 2.0: `python -c "import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"`

**Describe the current behavior**
It is not possible to train the "trainable" parameters of the RandomFourierFeatures keras layer, when using eager execution.

**Describe the expected behavior**
It should be possible to train the "trainable" parameters of the RandomFourierFeatures keras layer, even when using eager execution.

**Standalone code to reproduce the issue**
import tensorflow as tf
from tensorflow_core.python.keras.layers import RandomFourierFeatures

fourier_features = RandomFourierFeatures(
    1,
    kernel_initializer='gaussian',
    scale=1.0,
    trainable=True,
    dtype=tf.float64
)

input = tf.keras.Input(shape=(1,), dtype=tf.float64, name='input')
output = fourier_features(input)
model = tf.keras.Model(inputs=input, outputs=output)
model.compile(loss='mean_squared_error')

model.fit(tf.constant([[1.0]]), tf.constant([[1.0]]), epochs=1)

**Other info / logs**
The call to fit throws the following error:
ValueError: No gradients provided for any variable: ['random_fourier_features/random_features_scale:0'].
1/1 [==============================] - 0s 17ms/sample

### PR DESCRIPTION
**Pull request by johnamcleod**
fix #39088 

When executing in eager mode, computing the `kernel` value is done at build time, so it is not possible to calculate the gradient of the output with respect to the trainable parameters.

### CODE REVIEW

**Comment by gbaned** [comment]
@johnamcleod  Can you please resolve conflicts? Thanks!

**Comment by johnamcleod** [comment]
@gbaned - my apologies for the delay.

I have resolved the conflicts. It appears that the issue I was attempting to fix has been addressed since I opened this PR. As such this PR now only contains a new unit test to check that the RFF layer can be trained. I believe that this is still a valuable test to include in the test suite for this layer.

File: tensorflow/python/keras/layers/kernelized_test.py [hunk 1]
```diff
@@ -392,6 +393,19 @@ def test_good_kernel_approximation_multiple_inputs(self, initializer, scale,
     exact_kernel_matrix = exact_kernel_fn(x, y)
     self._assert_all_close(approx_kernel_matrix, exact_kernel_matrix, atol=0.05)
 
+  def testTrainLayer(self):
+    """Ensure the layer may be trained as part of a model."""
```
Comment (qlzh727) [inline tensorflow/python/keras/layers/kernelized_test.py:397]: You could use layer_test() which will do a bunch of extra verifications for a layer, eg saving/loading etc.

See https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/testing_utils.py#L87

**Comment by gbaned** [comment]
@johnamcleod Can you please check @qlzh727's comments and keep us posted ? Thanks!

File: tensorflow/python/keras/layers/kernelized_test.py (hunk 1 above)
Comment (johnamcleod) [reply tensorflow/python/keras/layers/kernelized_test.py:397]: Thanks for your suggestion - this looks really cool and is exactly what I need to use here. Unfortunately I am having some problems using this function with the layer - I am seeing errors that suggest the layer has not been registered. As soon as I can get this cleared up I will update the PR. My apologies for the delay.

Comment (gbaned) [reply tensorflow/python/keras/layers/kernelized_test.py:397]: @johnamcleod, Any update on this PR? Please. Thanks!

Comment (johnamcleod) [reply tensorflow/python/keras/layers/kernelized_test.py:397]: I was not able to figure out how to use the `layer_test` function with this layer, so I will close the PR. I apologise for the delay.

**Pull request closed without merging**